### ISSUE TITLE
[plugins] Update v2 embedding plugins to the latest stable release

### ISSUE THREAD

**Issue by mklim** [flutter/flutter#47153]
The v2 embedding plugins added support for v1.9, but v1.12 adds a few fixes that help the v2 path significantly. We should migrate the plugins to use v1.12 now that it's the most recent stable. 
For each plugin, we should:

1. Change the minimum supported Flutter version in the `pubspec.yaml` to `^v1.12.13+hotfix.5`. 
2. (Optional) Remove any constraint in the example's pubspec.yaml, if it exists.
3. Remove the `afterEvaluate` block hacking in inclusion of the lifecycle library.
4. Add the embedding v2 metadata xml tag. Otherwise, we're not autogenerating the v2 plugin registration. But since the tool can only autogenerate for one version of the embedding (even if it can toggle between v1 and v2 plugin APIs against the v2 embedding), we'd have to manually register the plugin in the EmbedderV1Activity.java.
5. Change to the getBinaryMessenger getter since getFlutterEngine.getDartExecutor is deprecated.
6. (Optional) we can remove the MainActivity for the v2 embedding in the example now since it has no content.
7. Bump the plugin version

These look like the plugins that need to be updated:

- [x] android_alarm_manager
- [x] android_intent (flutter/plugins#2586)
- [x] battery (flutter/plugins#2400)
- [x] camera
- [x] connectivity (flutter/plugins#2613)
- [x] device_info (flutter/plugins#2590)
- [x] e2e
- [ ] firebase_admob
- [ ] firebase_analytics
- [ ] firebase_auth
- [ ] firebase_core
- [ ] firebase_crashlytics
- [ ] firebase_database
- [ ] firebase_dynamic_links
- [ ] firebase_in_app_messaging
- [ ] firebase_messaging
- [ ] firebase_ml_vision
- [ ] firebase_performance
- [ ] firebase_remote_config
- [ ] firebase_storage
- [ ] flutter_plugin_android_lifecycle
- [x] google_sign_in
- [x] in_app_purchase
- [x] local_auth
- [x] package_info (flutter/plugins#2603)
- [x] path_provider
- [x] quick_actions (flutter/plugins#2605)
- [x] sensors
- [x] share
- [x] shared_preferences
- [x] url_launcher
- [x] video_player
- [x] webview_flutter (flutter/plugins#2425)

_ETA_: The current state of these plugins also print out deprecation warnings at build time because they access `getEngine()` directly.

**Comment by xster** [comment]
Did one for battery to see what it would look like https://github.com/flutter/plugins/pull/2400

**Comment by xster** [comment]
A few more actions to add to the list:

- Add the embedding v2 metadata xml tag. Otherwise, we're not autogenerating the v2 plugin registration. But since the tool can only autogenerate for one version of the embedding (even if it can toggle between v1 and v2 plugin APIs against the v2 embedding), we'd have to manually register the plugin in the EmbedderV1Activity.java. 
- Change to the getBinaryMessenger getter since getFlutterEngine.getDartExecutor is deprecated.

**Comment by mklim** [comment]
@xster thanks, added those steps to the original comment.

**Comment by ryanheise** [comment]
I would love to have some documentation on how to perform this migration, ideally with a prominent link to it from the 1.12 release notes. As a plugin author, it would help me to prepare my plugin for users who are already applying the new app project template (see https://github.com/ryanheise/audio_service/issues/138).

Part of the migration instructions could mention what to do with plugins like `android_alarm_manager` that do background execution. The old story required users of such a plugin to create their own custom `Application` class implementing a callback to register all plugins, which is called from the background context by the plugin, and I assume this has changed with the new approach (hopefully @bkonyi can also update the [Medium article](https://medium.com/flutter-io/executing-dart-in-the-background-with-flutter-plugins-and-geofencing-2b3e40a1a124) on background execution accordingly).

On a related point, it is not clear from just viewing commit histories and these GitHub issue discussions what command I need to run or what file I need to edit to cause the flutter tool to actually start generating the new `GeneratedPluginRegistrant` class for those who want to migrate an existing app project to use the new project structure, which I would also want to do for the example in my plugin.

**Comment by xster** [comment]
@ryanheise are the descriptions at https://flutter.dev/go/android-plugin-migration what you had in mind? We unfortunately didn't describe any specific to background tasks. Let us know if you couldn't figure it out from the migration guide.

**Comment by ryanheise** [comment]
@xster Thanks for the link. I will need some help specific to background tasks, though. I am using a technique similar to `android_alarm_manager` to register plugins in the background context but this plugin has not been migrated yet, and it seems neither of the two plugins in the above list that have been migrated showcase background tasks. The plugin migration guide doesn't actually explain any concepts behind the new model, such as how plugins actually get registered. I understood from @bkonyi 's Medium article that the activity and the service will essentially both need to independently register all plugins, and that is why plugin users needed to implement this custom Application class to facilitate in registering the plugins on behalf of the background task. Without any explanation of the new model, and without it being covered as an explicit step in the migration guide, and without any of the official plugins that use background tasks having yet been migrated to the new model, I feel it would be wise for me at this stage to request and wait for documentation :-)

As for some general feedback on the plugin migration guide that you linked, the document assumes that I know what V1 of embedding is. I haven't been paying attention to embedding (i.e. embedding Flutter into an existing app), and my reason for taking an interest now is not because I have suddenly become interested in the embedding use case, but rather that all plugin authors are now being asked to use the embedding API, even if the entire app is in Flutter. My question is, is V1 an earlier version of this embedding API? Or, is V1 simply "The old APIs based on PluginRegistry.Registrar" which the document refers to in the first sentence of that document? It may seem like a naive question, but for someone like myself who doesn't know the answer, this document keeps using the term V1 without providing any link to the page that describes what V1 is. If V1 is an earlier version of the embedding API, then I would be happy to skip this optional migration step. But if V1 is essentially the old plugin APIs, then I should probably do this migration step so that all plugin users with old project structures won't be inconvenienced.

The plugin migration guide also links to https://github.com/flutter/flutter/wiki/Upgrading-pre-1.12-Android-projects which almost addresses my last question on how to actually migrate an app. But that page doesn't explicitly say what step actually causes the new `GeneratedPluginRegistrant` class to be generated, and none of the listed steps seem to have an obvious connection to this.

**Comment by xster** [comment]
@ryanheise thanks for the feedback and sorry for the inconvenience.

re background, we updated the `android_alarm_manager` but indeed did not yet produce a set of official best practices for handling background execution in general yet. TBD in https://github.com/flutter/flutter/issues/32164.

The v1 embedding is indeed "The old APIs based on PluginRegistry.Registrar". @matthew-carroll can you see if there's anything else we can clarify in the migration wiki based on @ryanheise's comment? I added a small clarification to the GeneratedPluginRegistrant part.

**Comment by ryanheise** [comment]
Thanks, @xster, those clarifications do help, and also thanks for the pointers you gave on my project page. I can see we're on a path to a much nicer situation post v2, and pushing out this documentation could really help to mitigate a difficult transition period for both plugin developers and users, particularly users, since each single user depends on many plugins increasing their chances of running into a plugin like mine that is stuck on the migration path.

**Comment by ryanheise** [comment]
> re background, we updated the `android_alarm_manager`

It doesn't seem to be fully updated yet.

Currently the `MainActivity` class in the example looks like this:

```dart
public class MainActivity extends FlutterActivity {
  // TODO(bkonyi): Remove this once v2 of GeneratedPluginRegistrant rolls to stable. https://github.com/flutter/flutter/issues/42694
  @Override
  public void configureFlutterEngine(FlutterEngine flutterEngine) {
    ShimPluginRegistry shimPluginRegistry = new ShimPluginRegistry(flutterEngine);
    flutterEngine.getPlugins().add(new AndroidAlarmManagerPlugin());
    flutterEngine.getPlugins().add(new E2EPlugin());
    PathProviderPlugin.registerWith(
        shimPluginRegistry.registrarFor("io.flutter.plugins.pathprovider.PathProviderPlugin"));
  }
}
```

There is also an `Application` class which looks like this:

```dart
public class Application extends FlutterApplication implements PluginRegistrantCallback {
  @Override
  public void onCreate() {
    super.onCreate();
    AlarmService.setPluginRegistrant(this);
  }

  @Override
  public void registerWith(PluginRegistry registry) {
    GeneratedPluginRegistrant.registerWith(registry);
  }
}
```

The critical thing I was looking for in this project was how to initialise the background isolate.

There are various comments in `FlutterBackgroundExecutor` which touch on things that I would love to see fully explained in the guide and API documentation:

```java
      // We need to create an instance of `FlutterEngine` before looking up the
      // callback. If we don't, the callback cache won't be initialized and the
      // lookup will fail.
      FlutterCallbackInformation flutterCallback =
          FlutterCallbackInformation.lookupCallbackInformation(callbackHandle);
```

This isn't mentioned in the migration guide and is also not mentioned in the javadoc for `FlutterCallbackInformation`.

Also:

```java
      // The pluginRegistrantCallback should only be set in the V1 embedding as
      // plugin registration is done via reflection in the V2 embedding.
      if (pluginRegistrantCallback != null) {
        pluginRegistrantCallback.registerWith(new ShimPluginRegistry(backgroundFlutterEngine));
      }
```

The migration guide does not explain how this reflection approach works.

Assuming that for each `FlutterEngine` instance, the `GeneratedPluginRegistrant` is automatically called and a new instance of the plugin class is created and attached to the engine resulting in a 1-to-1 mapping between engines and plugins, how does a plugin instance detect whether it is running in a flutter engine that is associated with the activity/ui vs the headless engine? I used to be able to test whether `registrar.activity()` returned null, but now I assume I have to override `onAttachedToEngine` and figure this out from the `FlutterPluginBinding` parameter.

The `androidalarmmanager` plugin overrides this and uses `getFlutterEngine` which looks almost promising except that this is a deprecated API:

```java
  @Override
  public void onAttachedToEngine(FlutterPluginBinding binding) {
    onAttachedToEngine(
        binding.getApplicationContext(), binding.getFlutterEngine().getDartExecutor());
  }
```

And even if it were fine for me to call the deprecated `getFlutterEngime` method, I do not see any way from there to get the `RenderSurface` and test whether it's null.

**Comment by ryanheise** [comment]
I can now see that `ServiceAware` might be the way to do it, but this API doesn't seem to work, and none of the official plugins demonstrate how to use it. I made my plugin implement both `ActivityAware` and `ServiceAware` but only the lifecycle methods of the former were actually called. I also passed in my `Service` instance as the constructor argument when creating the background `FlutterEngine` but this itself didn't cause the `ServiceAware` callbacks to be called.

In any event, I got my plugin working with v2 using a different approach.

As some constructive feedback, I do appreciate all the hard work the Flutter team is doing, BUT I also hope that this migration experience is not something I will ever have to go through again. My suggestion is that a comprehensive migration guide should be in place before the release is rolled out, and should be prominently linked to from the release notes.

**Comment by mklim** [comment]
Thanks for the detailed feedback @ryanheise, I filed flutter/flutter#49578 for the docs lacking information about how to migrate plugins that execute in the background.

> It doesn't seem to be fully updated yet.

Yeah, sorry for the confusion here. That's what this issue is open to track. Most of the plugins (including `android_alarm_manager`) had v2 embedding support added before the API was completely rolled into the flutter stable channel of the time. Because of that the v2 support included compromises that we wouldn't recommend normally and now want to remove, the code snippets you point out being some of them. As far as I'm aware the only plugins that are fully updated to match our latest recommended practice on stable are the ones that have been checked off in the first comment in this issue.

> how does a plugin instance detect whether it is running in a flutter engine that is associated with the activity/ui vs the headless engine?

As far as I'm aware this should be the `ServiceAware` interface that you found. Since now the plugin is attached to the activity via a callback there's no upfront designation that the plugin is going to be always headless except for when it's constructed in a service and those callbacks are triggered instead. I think `ServiceAware` is only called if the Engine is constructed inside a service though, so that there are some cases where the Engine is in the background but constructed differently and it's not picked up? /cc @matthew-carroll and @bkonyi to confirm, I'm not sure about those cases specifically or what to do to recognize them. This sounds like it may be a bug on its own to me.

Good to hear you did get it working in the end.

**Comment by matthew-carroll** [comment]
@ryanheise I'll elaborate a bit on how plugins work in v2.

Any plugin, at any time, might be attached to a `FlutterEngine` that itself is attached to an `Activity`, `Service`, `ContentProvider`, or `BroadcastReceiver`. Or the `FlutterEngine` might not be attached to any of those other components - maybe the `FlutterEngine` is just sitting in a variable property somewhere and that's it. Moreover, a given `FlutterEngine` might detach from any of those components at any time.

Moving forward, every plugin needs to be able to handle the attachment to, and detachment from any Android component. The way that a plugin responds to attachment/detachment is by choosing to implement the associated `Aware` interface, e.g., `ActivityAware`, `ServiceAware`, etc. Those interfaces provide methods that are invoked when a `FlutterEngine` attaches to, or detaches from, one of the aforementioned Android components.

That said, many plugins won't even care if they're attached to a component or not. And those that do care, will usually only care about an `Activity`. That's why you will find many plugins that do not implement `ActivityAware` or `ServiceAware`. And only some implement `ActivityAware`.

That's the gist of how plugins operate in v2 with regard to Android components.

**Now, to address your question about `ServiceAware`...**

Whichever Android component is attached to a `FlutterEngine`, it is the job of that Android component to invoke the appropriate methods on the `FlutterEngine` to become attached. For example, here is some code that executes within `FlutterActivity` to attach the `Activity` to the `FlutterEngine`:

https://github.com/flutter/engine/blob/master/shell/platform/android/io/flutter/embedding/android/FlutterActivityAndFragmentDelegate.java#L180

https://github.com/flutter/engine/blob/master/shell/platform/android/io/flutter/embedding/android/FlutterActivityAndFragmentDelegate.java#L454

The reason that plugins have their `ActivityAware` callbacks invoked automatically within `FlutterActivity` is because Flutter has implemented that `Activity` to fulfill the `FlutterEngine`s contract. If you were to create your own `Activity` and host a `FlutterEngine`, you would become responsible for invoking those methods at the appropriate time.

So now consider a `Service`. Flutter does not currently offer a `FlutterService` (though I have considered creating one). Since there is no such thing as a `FlutterService`, you need to create your own `Service` definition. When you define your own `Service`, you need to invoke the appropriate methods on `FlutterEngine` to let it know that it's attached to your `Service`. For example, within your `Service` you need to do the following:

```
myFlutterEngine
  .getServiceControlSurface()
  .attachToService(this, getLifecycle(), /* is foreground */ false);
```

Then when your `Service` is destroyed, you need to invoke

```
myFlutterEngine
  .getServiceControlSurface()
  .detachFromService()
```

If you would like your `Service` to move from foreground to background, or visa versa, then also be sure to invoke the following at the appropriate times:

```
myFlutterEngine
  .getServiceControlSurface()
  .onMoveToForeground();

// and

myFlutterEngine
  .getServiceControlSurface()
  .onMoveToBackground();
```

Does this clarify your questions WRT to what you're experiencing?

I know @mklim has filed a ticket to create official documentation on this topic.

**Comment by ryanheise** [comment]
@matthew-carroll yes that clarifies the `ServiceAware` issue. FYI, I have to extend `MediaBrowserServiceCompat` so in my case the `FlutterService` approach wouldn't work for me. In any case, the above is helpful documentation and should be incorporated into various official places where people can find it.

This is very difficult to figure out based on the current documentation.

I also spoke to soon when I said I got v2 embedding working. Yes, it works, but it is also leaking memory (project: https://github.com/ryanheise/audio_service). Leak canary report:

```
====================================
HEAP ANALYSIS RESULT
====================================
1 APPLICATION LEAKS

References underlined with "~~~" are likely causes.
Learn more at https://squ.re/leaks.

3317 bytes retained by leaking objects
Signature: 8131736f850c542fc53bf276d4d1420433ae890
┬───
│ GC Root: Global variable in native code
│
├─ io.flutter.embedding.engine.FlutterJNI class
│    Leaking: NO (a class is never leaking)
│    ↓ static FlutterJNI.asyncWaitForVsyncDelegate
│                        ~~~~~~~~~~~~~~~~~~~~~~~~~
├─ io.flutter.view.VsyncWaiter$1 instance
│    Leaking: UNKNOWN
│    Anonymous class implementing io.flutter.embedding.engine.FlutterJNI$AsyncWaitForVsyncDelegate
│    ↓ VsyncWaiter$1.this$0
│                    ~~~~~~
├─ io.flutter.view.VsyncWaiter instance
│    Leaking: UNKNOWN
│    ↓ VsyncWaiter.windowManager
│                  ~~~~~~~~~~~~~
├─ android.view.WindowManagerImpl instance
│    Leaking: UNKNOWN
│    ↓ WindowManagerImpl.mContext
│                        ~~~~~~~~
├─ android.app.ContextImpl instance
│    Leaking: UNKNOWN
│    ↓ ContextImpl.mAutofillClient
│                  ~~~~~~~~~~~~~~~
╰→ com.ryanheise.audioserviceexample.MainActivity instance
​     Leaking: YES (ObjectWatcher was watching this because com.ryanheise.audioserviceexample.MainActivity received Activity#onDestroy() callback and Activity#mDestroyed is true)
​     key = e55cb0c6-64a7-4ff1-9cfa-dc5911938e58
​     watchDurationMillis = 5186
​     retainedDurationMillis = 180
====================================
0 LIBRARY LEAKS
```

Steps to reproduce:

1. Run the example
2. Press play
3. Press pause (maybe an optional step)
4. Use task manager to swipe away task. Now the activity is destroyed while the service is still running as intended.

So an instance of the main activity is being retained but you might need to be the designer of the v2 embedding architecture to be able to figure out where the leak is.

**Comment by matthew-carroll** [comment]
@ryanheise would you mind filing a separate issue for the memory leak concern? It's best if we only track one problem per issue.

**Comment by ryanheise** [comment]
@matthew-carroll No problem. Moved this one to #49612

**Comment by ened** [comment]
@mklim maybe could you add a link to https://flutter.dev/docs/development/packages-and-plugins/plugin-api-migration to the top, so that others can find the exact steps quickly?

**Comment by mklim** [comment]
@ened those instructions are intended for plugin developers that are adding v2 embedding support for the first time after the 1.12 stable release. I don't think this bug would be relevant for that since these steps all have to do with workarounds flutter/plugins put in place to add support pre 1.12. Could be missing something though, did you have something in mind?

**Comment by ened** [comment]
@mklim no, after checking https://github.com/flutter/plugins/pull/2590 it's all clear.

**Comment by xster** [comment]
All non-firebase ones are cleaned up in https://github.com/flutter/plugins/pull/2778

**Comment by xster** [comment]
Closing this bug and branching off firebase cleanups to https://github.com/FirebaseExtended/flutterfire/issues/2602.

**Comment by VLADPOLUNIN** [comment]
How are you feeling?
